File: tauri-svelte-preview/src/lib/shell/usage/usageCurrent.ts
import type { ProviderUsageSnapshot } from './usageTypes.ts';

type ProviderUsageWindowInput = {
  label?: string;
  name?: string;
  usedPercent?: number | null;
  percentConsumed?: number | null;
  percentRemaining?: number | null;
  resetsAt?: string | null;
  resetAt?: string | null;
  windowMinutes?: number | null;
  semantics?: string;
};

type ProviderUsageInput = Omit<Partial<ProviderUsageSnapshot>, 'windows'> & {
  windows?: ProviderUsageWindowInput[];
  inputTokens?: number;
};

export function usageWindowDurationLabel(windowMinutes: number | null | undefined): string {
  if (!Number.isFinite(windowMinutes) || !windowMinutes || windowMinutes <= 0) return 'Provider window';
  if (windowMinutes >= 6 * 24 * 60 && windowMinutes <= 8 * 24 * 60) return 'Weekly';
  if (windowMinutes < 60) return `${windowMinutes}-minute`;
  const hours = windowMinutes / 60;
  return `${Number.isInteger(hours) ? hours : hours.toFixed(1)}-hour`;
}

export function usageQuotaWindowLabel(window: { label: string; windowMinutes?: number | null }): string {
  const minutes = window.windowMinutes;
  if (minutes && minutes >= 6 * 24 * 60 && minutes <= 8 * 24 * 60) return 'Weekly (7-day) window';
  if (minutes === 5 * 60) return 'Session (5-hour) window';
  const label = window.label.trim() || 'Provider';
  return /window$/i.test(label) ? label : `${label} window`;
}

function reportedWindowMinutes(window: ProviderUsageWindowInput): number | null {
  if (Number.isFinite(window.windowMinutes) && window.windowMinutes && window.windowMinutes > 0) return window.windowMinutes;
  const semanticsMatch = window.semantics?.match(/\b(\d+)-minute\b/);
  return semanticsMatch ? Number(semanticsMatch[1]) : null;
}

function normalizeProviderWindow(window: ProviderUsageWindowInput) {
  const windowMinutes = reportedWindowMinutes(window);
  const usedPercent = window.usedPercent ?? window.percentConsumed ?? (window.percentRemaining == null ? 0 : 100 - window.percentRemaining);
  return {
    label: windowMinutes == null ? window.label?.trim() || window.name?.trim() || 'Provider window' : usageWindowDurationLabel(windowMinutes),
    usedPercent: Math.max(0, Math.min(100, usedPercent)),
    resetsAt: window.resetsAt ?? window.resetAt ?? null,
    windowMinutes
  };
}

export function normalizeProviderUsage(input: ProviderUsageInput): ProviderUsageSnapshot {
  const state = input.state === 'available' && (input.windows?.length ?? 0) > 0 ? 'available' : input.state === 'error' ? 'error' : 'unavailable';
  return {
    provider: input.provider?.trim() || 'unknown',
    account: input.account ?? null,
    instanceId: input.instanceId?.trim() || 'unknown',
    state,
    windows: state === 'available' ? (input.windows ?? []).map(normalizeProviderWindow) : [],
    capturedAt: input.capturedAt ?? Date.now(),
    source: input.source ?? null,
    sourceVersion: input.sourceVersion ?? null,
    unavailableReason: state === 'available' ? null : input.unavailableReason ?? 'No provider quota data is available.'
  };
}

export function usageWindowLabel(window: { label: string; usedPercent: number }): string {
  return `${window.label}: ${Math.round(window.usedPercent)}% used`;
}

export function usageProviderLabel(provider: string): string {
  return provider.trim() ? provider.trim().replace(/[-_]+/g, ' ').replace(/\b\w/g, (character) => character.toUpperCase()) : 'Provider';
}

export function usageResetLabel(resetsAt: string | null): string {
  if (!resetsAt) return 'Reset unavailable';
  const parsed = Number(resetsAt);
  const target = Number.isFinite(parsed) ? (parsed < 10_000_000_000 ? parsed * 1000 : parsed) : Date.parse(resetsAt);
  if (!Number.isFinite(target)) return 'Reset unavailable';
  const delta = Math.max(0, target - Date.now());
  const minutes = Math.round(delta / 60_000);
  if (minutes < 60) return `Resets in ${minutes}m`;
  const hours = Math.floor(minutes / 60);
  const remainder = minutes % 60;
  return `Resets in ${hours}h${remainder ? ` ${remainder}m` : ''}`;
}

export function usagePercent(window: { usedPercent: number }): number {
  return Math.max(0, Math.min(100, window.usedPercent));
}

File: tauri-svelte-preview/src-tauri/src/usage_current.rs
use crate::usage_sources::read_latest_local_quota;
use crate::{claude_quota, usage_remote};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ProviderUsageState {
    Available,
    Unavailable,
    Error,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ProviderUsageWindow {
    pub label: String,
    pub used_percent: f64,
    pub resets_at: Option<String>,
    pub window_minutes: Option<u64>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ProviderUsageSnapshot {
    pub provider: String,
    pub account: Option<String>,
    pub instance_id: String,
    pub state: ProviderUsageState,
    pub windows: Vec<ProviderUsageWindow>,
    pub captured_at: u128,
    pub source: Option<String>,
    pub source_version: Option<String>,
    pub unavailable_reason: Option<String>,
}

pub trait ProviderUsageReader: Send + Sync {
    fn read(&self, provider: &str, instance_id: &str) -> ProviderUsageSnapshot;
}

#[tauri::command]
pub async fn read_current_provider_usage(
    provider: Option<String>,
    instance_id: Option<String>,
) -> Result<ProviderUsageSnapshot, String> {
    let provider = provider.unwrap_or_else(|| "unknown".to_string());
    let instance_id = instance_id.unwrap_or_else(|| "unknown".to_string());
    if provider.eq_ignore_ascii_case("claude") {
        let snapshot = match usage_remote::read_claude_quota().await {
            Ok(quota) => ProviderUsageSnapshot {
                provider: "claude".to_string(),
                account: quota.account,
                instance_id,
                state: ProviderUsageState::Available,
                windows: quota
                    .windows
                    .into_iter()
                    .map(|window| ProviderUsageWindow {
                        label: window.label,
                        used_percent: window.used_percent,
                        resets_at: window.resets_at,
                        window_minutes: window.window_minutes,
                    })
                    .collect(),
                captured_at: quota.captured_at,
                source: Some(claude_quota::USAGE_SOURCE.to_string()),
                source_version: Some(claude_quota::USAGE_SOURCE_VERSION.to_string()),
                unavailable_reason: None,
            },
            Err(error) if error.is_unavailable() => {
                unavailable_snapshot(&provider, &instance_id, &error.reason())
            }
            Err(error) => error_snapshot(&provider, &instance_id, &error.reason()),
        };
        return Ok(normalize_provider_usage(snapshot));
    }
    let snapshot = read_latest_local_quota(&provider)
        .map(|quota| ProviderUsageSnapshot {
            provider: quota.provider,
            account: quota.account,
            instance_id: instance_id.clone(),
            state: ProviderUsageState::Available,
            windows: quota
                .windows
                .into_iter()
                .map(|window| ProviderUsageWindow {
                    label: window.name,
                    used_percent: window.percent_consumed,
                    resets_at: window.reset_at,
                    window_minutes: window
                        .semantics
                        .split_whitespace()
                        .find_map(|part| part.strip_suffix("-minute"))
                        .and_then(|minutes| minutes.parse::<u64>().ok()),
                })
                .collect(),
            captured_at: quota.captured_at,
            source: Some("local provider rate-limit records".to_string()),
            source_version: Some("local-v1".to_string()),
            unavailable_reason: None,
        })
        .unwrap_or_else(|| UnavailableProviderUsageReader.read(&provider, &instance_id));
    Ok(normalize_provider_usage(snapshot))
}

fn error_snapshot(provider: &str, instance_id: &str, reason: &str) -> ProviderUsageSnapshot {
    ProviderUsageSnapshot {
        provider: provider.trim().to_string(),
        account: None,
        instance_id: instance_id.trim().to_string(),
        state: ProviderUsageState::Error,
        windows: Vec::new(),
        captured_at: now_ms(),
        source: Some(claude_quota::USAGE_SOURCE.to_string()),
        source_version: Some(claude_quota::USAGE_SOURCE_VERSION.to_string()),
        unavailable_reason: Some(reason.to_string()),
    }
}

#[derive(Debug, Default, Clone, Copy)]
pub struct UnavailableProviderUsageReader;

impl ProviderUsageReader for UnavailableProviderUsageReader {
    fn read(&self, provider: &str, instance_id: &str) -> ProviderUsageSnapshot {
        unavailable_snapshot(
            provider,
            instance_id,
            "The provider has not advertised quota data",
        )
    }
}

pub fn unavailable_snapshot(
    provider: &str,
    instance_id: &str,
    reason: &str,
) -> ProviderUsageSnapshot {
    ProviderUsageSnapshot {
        provider: provider.trim().to_string(),
        account: None,
        instance_id: instance_id.trim().to_string(),
        state: ProviderUsageState::Unavailable,
        windows: Vec::new(),
        captured_at: now_ms(),
        source: None,
        source_version: None,
        unavailable_reason: Some(reason.to_string()),
    }
}

/// Accept only a provider-authored quota snapshot. Transcript token counts are
/// deliberately not an input to this function.
pub fn normalize_provider_usage(mut snapshot: ProviderUsageSnapshot) -> ProviderUsageSnapshot {
    if snapshot.provider.trim().is_empty() || snapshot.instance_id.trim().is_empty() {
        snapshot.state = ProviderUsageState::Unavailable;
        snapshot.windows.clear();
        snapshot.unavailable_reason = Some("Provider identity is incomplete".to_string());
        return snapshot;
    }
    if snapshot.state == ProviderUsageState::Available && snapshot.windows.is_empty() {
        snapshot.state = ProviderUsageState::Unavailable;
        snapshot.unavailable_reason = Some("The provider returned no quota windows".to_string());
    }
    for window in &mut snapshot.windows {
        window.used_percent = window.used_percent.clamp(0.0, 100.0);
        if window.label.trim().is_empty() {
            window.label = "Provider window".to_string();
        }
    }
    snapshot
}

fn now_ms() -> u128 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn usage_current_never_infers_quota_from_tokens() {
        let snapshot = normalize_provider_usage(ProviderUsageSnapshot {
            provider: "provider-a".into(),
            account: None,
            instance_id: "instance-a".into(),
            state: ProviderUsageState::Unavailable,
            windows: Vec::new(),
            captured_at: 1,
            source: None,
            source_version: None,
            unavailable_reason: Some("No quota data".into()),
        });
        assert_eq!(snapshot.state, ProviderUsageState::Unavailable);
        assert!(snapshot.windows.is_empty());
    }
}
